startup/40-xspress3.py

```diff
@@ -0,0 +1,424 @@
+    # channel6 = C(Xspress3Channel, 'C6_', channel_num=6)
+    # channel7 = C(Xspress3Channel, 'C7_', channel_num=7)
+    # channel8 = C(Xspress3Channel, 'C8_', channel_num=8)
+
+
+xs = QASXspress3Detector('XF:07BMB-ES{Xsp:1}:', name='xs')
+for n in [1, 2, 3, 4]:
+    getattr(xs, f'channel{n}').rois.read_attrs = ['roi{:02}'.format(j) for j in [1, 2, 3, 4]]
+xs.hdf5.num_extra_dims.put(0)
+xs.channel2.vis_enabled.put(1)
+xs.channel3.vis_enabled.put(1)
+xs.settings.num_channels.put(4)
+
+
+class XSFlyer:
+    def __init__(self, *, pb, di, pb_triggers, xs_dets, an_dets, motor):
+        """
+        The flyer based on a single encoder pizza-box and multiple xspress3 devices running a mono.
+
+        Parameters
+        ----------
+        pb : an encoder pizza-box orchestrating the experiment
+        di : a digital input to timestamp trigger pulses
+        pb_triggers : list
+            a list of names of the trigger signals (e.g., 'do1') from the pizza-box digital outputs.
+            It's possible to configure up to 4 triggers per pizza-box, triggering a corresponding xspress3 detector each.
+        xs_dets : list
+            a list of ophyd devices for the xspress3 detectors (up to 4 xspress3 detectors per 1 pizza-box)
+        an_dets : list
+            a list of analog detectors to be triggered
+        motor : a monochromator motor
+        """
+        self.name = f'{pb.name}-{"-".join([xs_det.name for xs_det in xs_dets])}-{"-".join([an_det.name for an_det in an_dets])}-{motor.name}-{self.__class__.__name__}'
+        self.pb = pb
+        self.di = di
+        self.pb_triggers = pb_triggers
+        self.xs_dets = xs_dets
+        self.an_dets = an_dets
+        self.motor = motor
+
+        # self.parent = None
+        self.num_points = {}
+        self._motor_status = None
+
+    def __repr__(self):
+        return f"""\
+    Flyer '{self.name}' with the following config:
+
+        - encoder pizza-box  : {self.pb.name}
+        - pizza-box triggers : {', '.join([x for x in self.pb_triggers])}
+        - xspress3 detectors : {', '.join([x.name for x in self.xs_dets])}
+        - analog detectors   : {', '.join([x.name for x in self.an_dets])}
+        - motor              : {self.motor.name}
+"""
+
+    def kickoff(self, *args, **kwargs):
+        # Set all required signals in xspress3
+        self._calc_num_points()
+        for xs_det in self.xs_dets:
+            xs_det.stage()
+            # TODO: set the "XF:07BMB-ES{Xsp:1}:HDF5:FileWriteMode" PV to "Stream" (option 2)
+            xs_det.settings.num_images.put(self.num_points[xs_det.name])
+            xs_det.hdf5.num_capture.put(self.num_points[xs_det.name])
+
+        # self.det.external_trig.put(True)  # This is questionable, if we need it or not.
+        # Next line should take care of everything.
+        # xspress3 CSS:
+        for xs_det in self.xs_dets:
+            xs_det.settings.trigger_mode.put(3)  # "Acquisition Controls and Status" (top left pane) -->
```

Comment:
We have a slight preference for using the enum strings in the code.  It is a bit less stable API wise, but much more readable.

Reply:
We implemented this. Thank you Tom.